File: backend/apps/authentication/interfaces/i_user_admin_actions.py
"""
IUserAdminActions — ISP interface for admin user management (HU-14, D25).

Responsibility (SRP): declare the admin-only user operations. Separate from
    IAuthService (session) so AuthService and UserAdminService don't mix concerns.
Pattern: ISP. SOLID: ISP · DIP · OCP.

D25: user management lives in apps/authentication/ (where User already is), NOT a new
    users/ module. UserAdminService implements this; views depend on the interface (DIP).
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from apps.authentication.models import User


class IUserAdminActions(ABC):

    @abstractmethod
    def list_users(self, filters: dict | None = None) -> list:
        """List users, optionally filtered by role/estado. Returns list of UserData dicts."""
        ...

    @abstractmethod
    def create_user(self, data: dict) -> dict:
        """Create a user with a given role (worker/admin). Returns the created UserData."""
        ...

    @abstractmethod
    def update_user(self, user_id: int, data: dict) -> dict:
        """Update only the managed user's first and/or last name."""
        ...

    @abstractmethod
    def block_user(self, user_id: int) -> dict:
        """Set estado = BLOQUEADO. Returns updated UserData."""
        ...

    @abstractmethod
    def unblock_user(self, user_id: int) -> dict:
        """Set estado = ACTIVO and reset failed attempts. Returns updated UserData."""
        ...

    @abstractmethod
    def retry_mailbox(self, user_id: int) -> dict:
        """Reintenta idempotentemente el buzón corporativo de un trabajador."""
        ...

    @abstractmethod
    def confirm_manual_mailbox(self, user_id: int, email: str, actor: User) -> dict:
        """Registra que el administrador creó el buzón manualmente en cPanel."""
        ...

    @abstractmethod
    def rotate_occupant(self, user_id: int, data: dict, actor: User) -> dict:
        """Cambia el ocupante y rota ambas credenciales sin alterar el correo."""
        ...

    @abstractmethod
    def rotate_occupant_manually(
        self,
        user_id: int,
        data: dict,
        actor: User,
    ) -> dict:
        """Cambia el ocupante tras confirmar la rotación manual del buzón."""
        ...

File: frontend/src/modules/auth/interfaces/IUserAdminActions.ts
/**
 * IUserAdminActions — FE contract for admin user management (HU-14, ISP).
 * Separate from IAuthService (session) — ISP. SOLID: ISP · DIP.
 */

export type MailboxStatus = 'creado' | 'pendiente' | 'no_aplica'
export type MailboxManagement = 'manual' | 'uapi' | 'no_aplica'

export interface AdminUser {
  id: string
  email: string
  nombre: string
  apellido: string
  rol: string
  estado: string
  emailVerificado: boolean
  buzonEstado: MailboxStatus
  buzonGestion: MailboxManagement
}

export interface AdminUserOperationResult extends AdminUser {
  appPassword?: string
  buzonPassword?: string
}

export interface CreateUserData {
  nombre: string
  apellido: string
  email: string
  password: string
  role: 'worker' | 'client'
}

export interface UpdateUserData {
  nombre?: string
  apellido?: string
}

export interface RotateOccupantData {
  nombre: string
  apellido: string
}

export interface RotateOccupantManualData extends RotateOccupantData {
  emailConfirmacion: string
  rotacionBuzonConfirmada: boolean
}

export interface IUserAdminActions {
  listUsers(filters?: { role?: string; estado?: string }): Promise<AdminUser[]>
  createUser(data: CreateUserData): Promise<AdminUserOperationResult>
  updateUser(id: string, data: UpdateUserData): Promise<AdminUser>
  blockUser(id: string): Promise<AdminUser>
  unblockUser(id: string): Promise<AdminUser>
  retryMailbox(id: string): Promise<AdminUserOperationResult>
  rotateOccupant(
    id: string,
    data: RotateOccupantData,
  ): Promise<AdminUserOperationResult>
  rotateOccupantManually(
    id: string,
    data: RotateOccupantManualData,
  ): Promise<AdminUserOperationResult>
}
